cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -43,11 +42,27 @@ type MasterConfiguration struct {
 type API struct {
 	AdvertiseAddresses []string
 	ExternalDNSNames   []string
-	BindPort           int32
+	Port               int32
 }
 
 type Discovery struct {
-	BindPort int32
+	HTTPS *HTTPSDiscovery
+	File  *FileDiscovery
+	Token *TokenDiscovery
+}
+
+type HTTPSDiscovery struct {
+	URL string
+}
+
+type FileDiscovery struct {
+	Path string
+}
+
+type TokenDiscovery struct {
+	TokenID   string
+	Token     string
```

Comment:
I do notice this leads to d.v.Token.Token etc, we could perhaps consider token redundant and go with Token.ID and Token.Secret.